File: PhoenixDemo-ObjectiveC/PhoenixDemo-ObjectiveC/AuthenticationViewController.m
//
//  AuthenticationViewController.m
//  PhoenixDemo-ObjectiveC
//
//  Created by Chris Nevin on 04/08/2015.
//  Copyright © 2015 Tigerspike. All rights reserved.
//

#import "AuthenticationViewController.h"

@import PhoenixSDK;

#import "PHXPhoenixManager.h"

typedef NS_ENUM(NSUInteger) {
    Login,
    LoggedIn,
    LoggingIn,
    LoginFailed,
} LoginMessage;

@interface AuthenticationViewController()
{
    LoginMessage currentStatus;
}
@end


@implementation AuthenticationViewController

- (LoginMessage)status {
    if ([self loggedIn]) {
        return LoggedIn;
    } else {
        return currentStatus;
    }
}

- (NSString*)messageForStatus {
    switch ([self status]) {
        case LoggedIn:
            return @"Logged in";
        case Login:
            return @"Login";
        case LoggingIn:
            return @"Logging in...";
        case LoginFailed:
            return @"Login failed!";
        default:
            return @"";
    }
}

- (UIColor*)colorForStatus {
    switch ([self status]) {
        case LoggingIn:
            return [UIColor purpleColor];
        case LoggedIn:
            return [UIColor grayColor];
        case LoginFailed:
            return [UIColor redColor];
        default:
            return [UIColor blackColor];
    }
}

- (Phoenix*)phoenix {
    return [[PHXPhoenixManager sharedManager] phoenix];
}

- (BOOL)loggedIn {
    return [[self phoenix] isLoggedIn];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setTitle:@"Authentication"];
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 2;
}

- (NSInteger)numberOfSectionsInTableView:(nonnull UITableView *)tableView {
    return 1;
}

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    if (indexPath.row == 0) {
        cell.textLabel.text = [self messageForStatus];
        cell.textLabel.textColor = [self colorForStatus];
        cell.userInteractionEnabled = !self.loggedIn;
    } else {
        cell.textLabel.text = @"Logout";
        cell.textLabel.textColor = !self.loggedIn ? [UIColor grayColor] : [UIColor blackColor];
        cell.userInteractionEnabled = self.loggedIn;
    }
    return cell;
}

- (void)tableView:(nonnull UITableView *)tableView didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:true];
    if (indexPath.row == 0) {
        [self login];
    } else {
        [self logout];
    }
}

- (void)login {
    
    if (self.loggedIn) { return; }
    
    currentStatus = LoggingIn;
    [self.tableView reloadData];
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Enter Details" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Username";
    }];
    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Password";
        textField.secureTextEntry = true;
    }];
    [alert addAction:[UIAlertAction actionWithTitle:@"Login" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        NSString *username = alert.textFields.firstObject.text;
        NSString *password = alert.textFields.lastObject.text;
        if (!(username.length != 0 && password.length != 0)) {
            currentStatus = Login;
            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                [self.tableView reloadData];
            }];
            return;
        }
        [self.phoenix loginWithUsername:username password:password callback:^(BOOL authenticated) {
            currentStatus = authenticated ? LoggedIn : LoginFailed;
            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                [self.tableView reloadData];
            }];
        }];
    }]];
    [self presentViewController:alert animated:true completion:nil];
}

- (void)logout {
    currentStatus = Login;
    [self.phoenix logout];
    [self.tableView reloadData];
}

@end

File: PhoenixDemo-Swift/PhoenixDemo-Swift/AuthenticationViewController.swift
//
//  AuthenticationViewController.swift
//  PhoenixDemo-Swift
//
//  Created by Chris Nevin on 04/08/2015.
//  Copyright © 2015 Tigerspike. All rights reserved.
//

import UIKit
import PhoenixSDK

// Valid messages
private enum LoginMessages: String {
    case None = ""
    case LoggedIn = "Logged in"
    case LoggingIn = "Logging in..."
    case Login = "Login"
    case LoginFailed = "Login Failed"
    
    func color() -> UIColor {
        switch self {
        case .LoggedIn:
            return .grayColor()
            
        case .LoginFailed:
            return .redColor()
            
        case .LoggingIn:
            return .purpleColor()
            
        default:
            return .blackColor()
        }
    }
}

class AuthenticationViewController: UITableViewController {

    // Store your user login status. In this demo we just use a static variable for it.
    private static var loggedInUser: Phoenix.User?
    private static var loginMessage: LoginMessages = .Login

    private var loggedIn:Bool {
        return AuthenticationViewController.loginMessage == .LoggedIn
    }
    
    private let ViewUserSegue = "LoginViewUser"
    
    override func viewDidLoad() {
        super.viewDidLoad()
        title = "Authentication"
        tableView.reloadData()
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == ViewUserSegue {
            if let viewUser = segue.destinationViewController as? ViewUserViewController {
                viewUser.user = AuthenticationViewController.loggedInUser
            }
        }
    }
    
    // MARK:- UITableViewDataSource
    
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 2
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath: indexPath)
        
        if indexPath.row == 0 {
            cell.textLabel?.text = AuthenticationViewController.loginMessage.rawValue
            cell.textLabel?.textColor = AuthenticationViewController.loginMessage.color()
            cell.userInteractionEnabled = !loggedIn
        }
        else {
            cell.textLabel?.text = "Logout"
            cell.userInteractionEnabled = loggedIn
            cell.textLabel?.textColor = loggedIn ? .blackColor() : .grayColor()
        }
        
        return cell
    }
    
    // MARK:- UITableViewDelegate
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        if indexPath.row == 0 {
            login()
        } else {
            logout()
        }
    }
    
    // MARK:- Log in/out
    
    func login() {
        if loggedIn {
            return
        }

        func reloadUI(message: LoginMessages) {
            AuthenticationViewController.loginMessage = message
            if NSThread.isMainThread() {
                tableView.reloadData()
            }
            else {
                NSOperationQueue.mainQueue().addOperationWithBlock() { [weak self] in
                    self?.tableView.reloadData()
                }
            }
        }
        reloadUI(.LoggingIn)
        
        let alert = UIAlertController(title: "Enter Details", message: nil, preferredStyle: UIAlertControllerStyle.Alert)
        
        alert.addTextFieldWithConfigurationHandler { (textField) -> Void in
            textField.placeholder = "Username"
        }
        
        alert.addTextFieldWithConfigurationHandler { (textField) -> Void in
            textField.placeholder = "Password"
            textField.secureTextEntry = true
        }
        
        
        alert.addAction(UIAlertAction(title: "Login", style: UIAlertActionStyle.Default, handler: { [weak self] (action) -> Void in
            guard let username = alert.textFields?.first?.text, password = alert.textFields?.last?.text else {
                reloadUI(.Login)
                return
            }
            
            PhoenixManager.phoenix.identity.login(withUsername: username, password: password, callback: { (user, error) -> () in
                reloadUI(error == nil ? .LoggedIn : .LoginFailed)
                
                if AuthenticationViewController.loginMessage == .LoggedIn {
                    AuthenticationViewController.loggedInUser = user
                    
                    NSOperationQueue.mainQueue().addOperationWithBlock() { [weak self] in
                        self?.performSegueWithIdentifier(ViewUserSegue, sender: self)
                    }
                }
            })
        }))
        
        alert.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: nil))
        
        presentViewController(alert, animated: true) { }
    }
    
    func logout() {
        AuthenticationViewController.loginMessage = .Login
        PhoenixManager.phoenix.identity.logout()
        tableView.reloadData()
    }
}